build.assets/tooling/cmd/build-apt-repos/main.go

```diff
@@ -0,0 +1,151 @@
+package main
+
+import (
+	"container/list"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/seqsense/s3sync"
+)
+
+func main() {
+	// artifactPath := "/artifacts"
+	// bucketName := "aptly-test20220223185137234100000001"
+	// localBucketPath := "/bucket"
+	// majorVersion := "vTODO"
+	// releaseChannel := "stable"
+	supportedOSs := map[string][]string{
+		"debian": {
+			"stretch",
+			"buster",
+			"bullseye",
+			"bookwork",
+			"trixie",
+		},
+		"ubuntu": {
+			"xenial",
+			"yakkety",
+			"zesty",
+			"artful",
+			"bionic",
+			"cosmic",
+			"disco",
+			"eoan",
+			"focal",
+			"groovy",
+		},
+	}
+
+	args := parseFlags()
+
+	awsSession := session.Must(session.NewSession())
+	aptly := SetupAptly()
+
+	downloadExistingRepo(awsSession, *args.bucketName, *args.localBucketPath)
+
+	repos := list.New()
+
+	// Recreate existing repos with existin debs
+	for repo := range aptly.CreateReposFromBucket(*args.localBucketPath) {
+		repoPath := filepath.Join(*args.localBucketPath, repo.PublishedRepoRelativePath())
+		aptly.ImportDeb(repo.Name(), repoPath)
+		repos.PushBack(repo)
+	}
+
+	// Create new repos that may be missing in the case we add more channels/OS specific support/major versions
+	for repo := range aptly.CreateReposFromArtifactRequirements(supportedOSs, *args.releaseChannel, *args.majorVersion) {
+		repos.PushBack(repo)
+	}
+
+	// Walk over the artifact filesystem tree and find debs
+	err := filepath.WalkDir(*args.artifactPath,
+		func(debPath string, d fs.DirEntry, err error) error {
+			fmt.Println(debPath)
+			if err != nil {
+				return err
+			}
+
+			if d.IsDir() {
+				return nil
+			}
+
+			fileName := d.Name()
+			if filepath.Ext(fileName) != ".deb" {
+				return nil
+			}
+
+			// Import new artifacts into all repos that match the artifact's requirements
+			for e := repos.Front(); e != nil; e = e.Next() {
+				repo := e.Value.(*repo)
+
+				// Other checks could be added here to ensure that a given deb gets added to the correct repo
+				// such as name or parent directory, facilitating os-specific artifacts
+				if repo.majorVersion != *args.majorVersion || repo.releaseChannel != *args.releaseChannel {
+					continue
+				}
+
+				fmt.Println("importing deb")
+				aptly.ImportDeb(repo.Name(), debPath)
+			}
+
+			return nil
+		},
+	)
+	if err != nil {
+		handleFatalError(err, "Failed to find and import debs")
+	}
+
+	// Build a map keyed by os info with value of all repos that support the os in the key
+	// This will be used to structure the publish command
+	categorizedRepos := make(map[string]*list.List)
+	for e := repos.Front(); e != nil; e = e.Next() {
+		r := e.Value.(*repo)
+
+		if osRepos, ok := categorizedRepos[r.OSInfo()]; ok {
+			osRepos.PushBack(r)
+		} else {
+			a := list.New()
+			a.PushBack(r)
+			categorizedRepos[r.OSInfo()] = a
+		}
+	}
+
+	for os, osRepoList := range categorizedRepos {
```

Comment:
Assuming the type changes suggested above, 
```go
for os, osRepos := range categorizedRepos {
    for i, r := range osRepos {
	log.Printf("%d %s", i, r.osVersion)        
    }
    aptly.PublishRepos(osRepos)
}
```

Reply:
I think I've resolved this here: https://github.com/gravitational/teleport/blob/rfd/0058-package-distribution/build.assets/tooling/cmd/build-apt-repos/apt_repo_tool.go#L90-L99